tsl/src/compression/compression.c

```diff
@@ -2340,66 +2342,97 @@ decompress_batches(RowDecompressor *decompressor, ScanKeyData *scankeys, int num
+	Chunk *comp_chunk;
 	RowDecompressor decompressor;
 	SegmentFilter *filter;
 
-	if (predicates)
-		fill_predicate_context(linitial(chunks), predicates, &filters, &is_null);
-	foreach (ch, chunks)
-	{
-		chunk = (Chunk *) lfirst(ch);
-		bool chunk_status_changed = false;
-		ScanKeyData *scankeys = NULL;
-		Bitmapset *null_columns = NULL;
-		int num_scankeys = 0;
-
-		chunk_rel = table_open(chunk->table_id, RowExclusiveLock);
-		comp_chunk = ts_chunk_get_by_id(chunk->fd.compressed_chunk_id, true);
-		comp_chunk_rel = table_open(comp_chunk->table_id, RowExclusiveLock);
-		decompressor = build_decompressor(comp_chunk_rel, chunk_rel);
-
-		if (filters)
-		{
-			scankeys =
-				build_update_delete_scankeys(&decompressor, filters, &num_scankeys, &null_columns);
-		}
-		if (decompress_batches(&decompressor,
-							   scankeys,
-							   num_scankeys,
-							   null_columns,
-							   is_null,
-							   &chunk_status_changed))
-		{
-			/*
-			 * tuples from compressed chunk has been decompressed and moved
-			 * to staging area, thus mark this chunk as partially compressed
-			 */
-			if (chunk_status_changed == true)
-				ts_chunk_set_partial(lfirst(ch));
-		}
+	bool chunk_status_changed = false;
+	ScanKeyData *scankeys = NULL;
+	Bitmapset *null_columns = NULL;
+	int num_scankeys = 0;
 
-		ts_catalog_close_indexes(decompressor.indexstate);
-		FreeBulkInsertState(decompressor.bistate);
+	fill_predicate_context(chunk, predicates, &filters, &is_null);
 
-		table_close(chunk_rel, NoLock);
-		table_close(comp_chunk_rel, NoLock);
+	chunk_rel = table_open(chunk->table_id, RowExclusiveLock);
+	comp_chunk = ts_chunk_get_by_id(chunk->fd.compressed_chunk_id, true);
+	comp_chunk_rel = table_open(comp_chunk->table_id, RowExclusiveLock);
+	decompressor = build_decompressor(comp_chunk_rel, chunk_rel);
+
+	if (filters)
+	{
+		scankeys =
+			build_update_delete_scankeys(&decompressor, filters, &num_scankeys, &null_columns);
+	}
+	if (decompress_batches(&decompressor,
+						   scankeys,
+						   num_scankeys,
+						   null_columns,
+						   is_null,
+						   &chunk_status_changed))
+	{
+		/*
+		 * tuples from compressed chunk has been decompressed and moved
+		 * to staging area, thus mark this chunk as partially compressed
+		 */
+		if (chunk_status_changed == true)
+			ts_chunk_set_partial(chunk);
 	}
 
+	ts_catalog_close_indexes(decompressor.indexstate);
+	FreeBulkInsertState(decompressor.bistate);
+
+	table_close(chunk_rel, NoLock);
+	table_close(comp_chunk_rel, NoLock);
+
 	foreach (lc, filters)
 	{
 		filter = lfirst(lc);
 		pfree(filter);
 	}
 }
+
+/*
+ * Traverse the plan tree to look for Scan nodes on uncompressed chunks.
+ * Once Scan node is found check if chunk is compressed, if so then
+ * decompress those segments which match the filter conditions if present.
+ */
+bool
+decompress_target_segments(PlanState *ps)
+{
+	RangeTblEntry *rte = NULL;
+	Chunk *current_chunk;
+	if (ps == NULL || ts_guc_enable_transparent_decompression == false)
```

Comment:
I don't think we can use this GUC here. We should add a separate one for DML decompression which blocks UPDATE/DELETE and ON CONFLICT operations just like it is in the current release.

New GUC is left for separate PR.

Reply:
Fixed.